django/db/models/sql/compiler.py

```diff
@@ -583,50 +583,28 @@ def get_combinator_sql(self, combinator, all):
                     raise DatabaseError(
                         "ORDER BY not allowed in subqueries of compound statements."
                     )
-        elif self.query.is_sliced and combinator == "union":
-            for compiler in compilers:
-                # A sliced union cannot have its parts elided as some of them
-                # might be sliced as well and in the event where only a single
-                # part produces a non-empty resultset it might be impossible to
-                # generate valid SQL.
-                compiler.elide_empty = False
-        parts = ()
-        selected = self.query.selected
+        parts = []
+        empty_compiler = None
         for compiler in compilers:
             try:
-                # If the columns list is limited, then all combined queries
-                # must have the same columns list. Set the selects defined on
-                # the query on all combined queries, if not already set.
-                if selected is not None and compiler.query.selected is None:
-                    compiler.query = compiler.query.clone()
-                    compiler.query.set_values(selected)
-                part_sql, part_args = compiler.as_sql(with_col_aliases=True)
-                if compiler.query.combinator:
-                    # Wrap in a subquery if wrapping in parentheses isn't
-                    # supported.
-                    if not features.supports_parentheses_in_compound:
-                        part_sql = "SELECT * FROM ({})".format(part_sql)
-                    # Add parentheses when combining with compound query if not
-                    # already added for all compound queries.
-                    elif (
-                        self.query.subquery
-                        or not features.supports_slicing_ordering_in_compound
-                    ):
-                        part_sql = "({})".format(part_sql)
-                elif (
-                    self.query.subquery
-                    and features.supports_slicing_ordering_in_compound
-                ):
-                    part_sql = "({})".format(part_sql)
-                parts += ((part_sql, part_args),)
+                parts.append(self._get_combinator_part_sql(compiler))
             except EmptyResultSet:
                 # Omit the empty queryset with UNION and with DIFFERENCE if the
                 # first queryset is nonempty.
                 if combinator == "union" or (combinator == "difference" and parts):
+                    empty_compiler = compiler
                     continue
                 raise
         if not parts:
             raise EmptyResultSet
+        elif len(parts) == 1 and combinator == "union" and self.query.is_sliced:
```

Comment:
The check could be even more rigid by only taking that branch if the single non-empty compiled query was also sliced but that didn't seem worth the complexity.

Reply:
I can remove this elif block (and the `empty_compiler` logic) without something failing (on SQLite)
>          # in the event the later is also sliced it might result in invalid
>          # SQL due to the usage of multiple LIMIT clauses.

Can we add a test that covers this scenario?

Reply:
This is a Postgres specific issues as detailed in ticket-34125 which added tests in b42bd1a193a49f4e7ede207d2c40898a3171fd41.

Try running `queries.test_qs_combinators.QuerySetSetOperationTests` tests against Postgres

```
Traceback (most recent call last):
  File "/Users/charettes/Workspace/django/django/db/backends/utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/charettes/Workspace/django/.venv/lib/python3.12/site-packages/psycopg/cursor.py", line 97, in execute
    raise ex.with_traceback(None)
psycopg.errors.SyntaxError: multiple LIMIT clauses not allowed
LINE 1: ...s_number" WHERE "queries_number"."num" <= 0 LIMIT 1) LIMIT 1
                                                                      ^
```

Reply:
Got it 🥳 Sorry I missed that, thank you 👍